domain/storage/service/storage.go

```diff
@@ -0,0 +1,145 @@
+	// GetModelDetails returns the model and controller UUID for the current model.
+	GetModelDetails() (storage.ModelDetails, error)
+	// ImportFilesystem associates a filesystem (either native or volume backed) hosted by a cloud provider
+	// with a new storage instance (and storage pool) in a model.
+	ImportFilesystem(ctx context.Context, name internalstorage.Name,
+		filesystem storage.FilesystemInfo) (internalstorage.ID, error)
+}
+
+// StorageService defines a service for storage related behaviour.
+type StorageService struct {
+	st             State
+	logger         logger.Logger
+	registryGetter corestorage.ModelStorageRegistryGetter
+}
+
+// ImportFilesystem associates a filesystem (either native or volume backed) hosted by a cloud provider
+// with a new storage instance (and storage pool) in a model.
+// The following error types can be expected:
+// - [coreerrors.NotSupported]: when the importing the kind of storage is not supported by the provider.
+// - [storageerrors.InvalidPoolNameError]: when the supplied pool name is invalid.
+func (s *StorageService) ImportFilesystem(
+	ctx context.Context, credentialInvalidatorGetter envcontext.ModelCredentialInvalidatorGetter, arg ImportStorageParams,
+) (internalstorage.ID, error) {
+	if arg.Kind != internalstorage.StorageKindFilesystem {
+		// TODO(axw) implement support for volumes.
+		return "", errors.Errorf("storage kind %q not supported", arg.Kind.String()).Add(coreerrors.NotSupported)
+	}
+	if !internalstorage.IsValidPoolName(arg.Pool) {
+		return "", errors.Errorf("pool name %q not valid", arg.Pool).Add(storageerrors.InvalidPoolNameError)
+	}
+
+	poolDetails, err := s.st.GetStoragePoolByName(ctx, arg.Pool)
+	if errors.Is(err, storageerrors.PoolNotFoundError) {
+		poolDetails = storage.StoragePoolDetails{
+			Name:     arg.Pool,
+			Provider: arg.Pool,
+			Attrs:    map[string]string{},
+		}
+	} else if err != nil {
+		return "", errors.Capture(err)
+	}
+	registry, err := s.registryGetter.GetStorageRegistry(ctx)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	provider, err := registry.StorageProvider(internalstorage.ProviderType(poolDetails.Provider))
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	details, err := s.st.GetModelDetails()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	resourceTags := map[string]string{
+		tags.JujuModel:      details.ModelUUID,
+		tags.JujuController: details.ControllerUUID,
+	}
+	filesystemInfo := storage.FilesystemInfo{Pool: arg.Pool}
+
+	// If the storage provider supports filesystems, import the filesystem,
+	// otherwise import a volume which will back a filesystem.
+	invalidatorFunc, err := credentialInvalidatorGetter()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	callCtx := envcontext.WithCredentialInvalidator(ctx, invalidatorFunc)
+
+	var attr map[string]any
+	if len(poolDetails.Attrs) > 0 {
+		attr = transform.Map(poolDetails.Attrs, func(k, v string) (string, any) { return k, v })
+	}
```

Comment:
Just remove the if, it'll be a no-op in the end.

Reply:
`transform.Map` returns an empty map when the input is nil which could be considered "bad" behaviour
YMMV. I wish `transform.Map` didn't do that.

in mongo land, Attrs was stored as `map[string]any` so no transformation was needed and tests in particular could assume Attrs was nil if the input config had nil Attrs. Without the check all the tests would need to transform the struct under test to have an empty map.